provider/input_event.py

```diff
@@ -0,0 +1,575 @@
+import os
+import json
+import logging
+try:
+    from queue import Queue
+except ImportError:
+    from Queue import Queue
+
+from virttest import data_dir
+from virttest import utils_misc
+
+
+DEP_DIR = data_dir.get_deps_dir('input_event')
+
+
+class AgentMessageType:
+
+    '''Agent message types.'''
+
+    SYNC = 'SYNC'
+    INFO = 'INFO'
+    READY = 'READY'
+    EVENT = 'EVENT'
+    ERROR = 'ERROR'
+
+
+class AgentState:
+
+    '''Agent state codes.'''
+
+    STOPPED = 0
+    GREETING = 1
+    LISTENING = 2
+
+
+EventTypeKey = 'type'
+DevNameKey = 'device'
+
+
+class EventType:
+
+    '''Event types.'''
+
+    KEYDOWN = 'KEYDOWN'
+    KEYUP = 'KEYUP'
+    POINTERMOVE = 'POINTERMOVE'
+    WHEELFORWARD = 'WHEELFORWARD'
+    WHEELBACKWARD = 'WHEELBACKWARD'
+    UNKNOWN = 'UNKNOWN'
+
+
+class KeyEventData:
+
+    '''Key event schema.'''
+
+    KEYCODE = 'keyCode'
+    SCANCODE = 'scanCode'
+
+
+class PointerEventData:
+
+    '''Pointer event schema.'''
+
+    XPOS = 'xPos'
+    YPOS = 'yPos'
+    ABS = 'abs'
+
+
+class WheelEventData:
+
+    '''Wheel event schema.'''
+
+    HSCROLL = 'hScroll'     # horizontal scroll wheel
+    ABS = 'abs'
+
+
+class _EventListener(object):
+
+    '''Base implementation for the event listener class.'''
+
+    agent_source = ''
+    agent_target = ''
+    python_bin = ''
+
+    def __init__(self, vm):
+        '''
+        Initialize the event listener.
+
+        :param vm: VM object.
+        '''
+        self.events = Queue()
+        self.targets = {}
+        self._vm = vm
+        self._ctrl_sh = vm.wait_for_login()
+        self._agent_sh = None
+        self._agent_state = AgentState.STOPPED
+        self._install()
+        self._listen()
+
+    def _install(self):
+        '''Install (copy) the agent into VM.'''
+        if not os.path.exists(self.agent_source):
+            raise IOError('agent program is missing')
+        self._vm.copy_files_to(self.agent_source, self.agent_target)
+
+    def _uninstall(self):
+        '''Uninstall the agent.'''
+        raise NotImplementedError()
+
+    def _launch(self):
+        '''Launch the agent.'''
+        self._agent_sh = self._vm.wait_for_login()
+        self._agent_sh.set_output_func(self._parse_output)
+        self._agent_sh.set_output_params(tuple())
+        cmd = ' '.join((self.python_bin, self.agent_target))
+        self._agent_sh.sendline(cmd)
+
+    def _terminate(self):
+        '''Terminate the agent.'''
+        self._agent_sh.sendcontrol('c')
+        self._agent_sh.close()
+        # workaround it here
```

Comment:
Please give a more clear comment for that issue, suggest:
```python
# session objects wants `output_func` to be serializable, but it does not make sense
# currently, drop the value to avoid the framework complainting about that
```

Reply:
thanks, has updated~